src/DurableTask.Netherite/Util/PartitionTimeout.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace DurableTask.Netherite
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A utility class for terminating the partition if some task takes too long.
+    /// Implemented as a disposable, with <see cref="IDisposable.Dispose"> used to cancel the timeout.
+    /// </summary>
+    class PartitionTimeout : IDisposable
+    {
+        readonly CancellationTokenSource tokenSource;
+        readonly Task timeoutTask;
+
+        public PartitionTimeout(IPartitionErrorHandler errorHandler, string task, TimeSpan timeout)
+        {
+            this.tokenSource = new CancellationTokenSource();
+            this.timeoutTask = Task.Delay(timeout, this.tokenSource.Token);
+
+            // if the timeout tasks runs to completion without being cancelled, terminate the partition
+            this.timeoutTask.ContinueWith(
+                 _ => errorHandler.HandleError(
+                     $"{nameof(PartitionTimeout)}",
+                     $"{task} timed out after {timeout}",
+                     e: null,
+                     terminatePartition: true,
+                     reportAsWarning: false),
+                 TaskContinuationOptions.OnlyOnRanToCompletion);
+        }
+
+        public void Dispose()
+        {      
+            // cancel the timeout task (if it has not already completed)
+            this.tokenSource.Cancel();
+
+            // dispose the token source after the timeout task has completed
+            this.timeoutTask.ContinueWith(_ => this.tokenSource.Dispose());
+        }
+    }
```

Comment:
Isn't there a race condition here? What if the startup completes just a millisecond before the timeout task. If so, it's possible startup to complete and the timeout task to "continueWith" terminating the partition right after.

I suppose that may be 'ok' in the sense that if the startup takes 9.999 minutes, it might as well have timed out.

Reply:
It is true that that is possible.

Like you say, I think it is o.k. in this case.

Also, by design of the error handler, it doesn't make much of a difference whether the partition is terminated immediately before startup finishes or immediately after startup finishes.

Reply:
That said, I found a way to fix this which is slightly cleaner.

Reply:
I think I found a way to fix this so I will do that.